test/test_qr.lua

```diff
@@ -0,0 +1,254 @@
+    local a = tensorFunc{{1, 2, 3}, {4, 5, 6}, {7, 8, 9}}
+    local expected_q = tensorFunc{
+        {-1.230914909793328e-01,  9.045340337332914e-01,
+         4.082482904638621e-01},
+        {-4.923659639173310e-01,  3.015113445777629e-01,
+         -8.164965809277264e-01},
+        {-8.616404368553292e-01, -3.015113445777631e-01,
+         4.082482904638634e-01},
+    }
+    local expected_r = tensorFunc{
+        {-8.124038404635959e+00, -9.601136296387955e+00,
+         -1.107823418813995e+01},
+        { 0.000000000000000e+00,  9.045340337332926e-01,
+         1.809068067466585e+00},
+        { 0.000000000000000e+00,  0.000000000000000e+00,
+         -8.881784197001252e-16},
+    }
+    checkQRWithExpected(testOpts, a,  expected_q, expected_r)
+  end
+end, {doubleTensorOnly=true})
+
+-- Decomposing a specific (wide) rectangular matrix.
+addTestVariations(tests, 'qrRectFat', function(testOpts)
+  -- The matrix is chosen to be full-rank.
+  local a = testOpts.tensorFunc{
+      {1,  2,  3,  4},
+      {5,  6,  7,  8},
+      {9, 10, 11, 13}
+  }
+  local expected_q = testOpts.tensorFunc{
+      {-0.0966736489045663,  0.907737593658436 ,  0.4082482904638653},
+      {-0.4833682445228317,  0.3157348151855452, -0.8164965809277254},
+      {-0.870062840141097 , -0.2762679632873518,  0.4082482904638621}
+  }
+  local expected_r = testOpts.tensorFunc{
+      { -1.0344080432788603e+01,  -1.1794185166357092e+01,
+        -1.3244289899925587e+01,  -1.5564457473635180e+01},
+      {  0.0000000000000000e+00,   9.4720444555662542e-01,
+         1.8944088911132546e+00,   2.5653453733825331e+00},
+      {  0.0000000000000000e+00,   0.0000000000000000e+00,
+         1.5543122344752192e-15,   4.0824829046386757e-01}
+  }
+  checkQRWithExpected(testOpts, a, expected_q, expected_r)
+end, {doubleTensorOnly=true})
+
+-- Decomposing a specific (thin) rectangular matrix.
+addTestVariations(tests, 'qrRectThin', function(testOpts)
+  -- The matrix is chosen to be full-rank.
+  local a = testOpts.tensorFunc{
+      { 1,  2,  3},
+      { 4,  5,  6},
+      { 7,  8,  9},
+      {10, 11, 13},
+  }
+  local expected_q = testOpts.tensorFunc{
+      {-0.0776150525706334, -0.833052161400748 ,  0.3651483716701106},
+      {-0.3104602102825332, -0.4512365874254053, -0.1825741858350556},
+      {-0.5433053679944331, -0.0694210134500621, -0.7302967433402217},
+      {-0.7761505257063329,  0.3123945605252804,  0.5477225575051663}
+  }
+  local expected_r = testOpts.tensorFunc{
+      {-12.8840987267251261, -14.5916298832790581, -17.0753115655393231},
+      {  0,                  -1.0413152017509357,  -1.770235842976589 },
+      {  0,                   0,                    0.5477225575051664}
+  }
+  checkQRWithExpected(testOpts, a, expected_q, expected_r)
+end, {doubleTensorOnly=true})
+
+-- Decomposing a sequence of medium-sized random matrices.
+addTestVariations(tests, 'randomMediumQR', function(testOpts)
+  torch.manualSeed(1)
```

Comment:
if you want to control the seed for this test, can you restore the seed after? we might lose randomized tests if this test executes first.